forc-plugins/forc-crypto/src/keys/vanity.rs

```diff
@@ -0,0 +1,440 @@
+use anyhow::{anyhow, Result};
+use fuel_crypto::{fuel_types::Address, PublicKey, SecretKey};
+use fuels_accounts::wallet::{generate_mnemonic_phrase, DEFAULT_DERIVATION_PATH_PREFIX};
+use fuels_core::types::bech32::{Bech32Address, FUEL_BECH32_HRP};
+use rayon::iter::{self, Either, ParallelIterator};
+use regex::Regex;
+use serde_json::json;
+use std::{fs, path::PathBuf, time::Instant};
+
+forc_util::cli_examples! {
+    crate::Command {
+        [ Generate a checksummed vanity address with a given prefix => "forc crypto vanity --starts-with aaa" ]
+    }
+}
+
+#[derive(Debug, clap::Parser)]
+#[clap(
+    version,
+    about = "Generate a vanity address",
+    after_help = "Generate vanity addresses for the Fuel blockchain"
+)]
+pub struct Arg {
+    /// Prefix regex pattern or hex string.
+    #[arg(long, value_name = "PATTERN", required_unless_present = "ends_with")]
+    pub starts_with: Option<String>,
+
+    /// Suffix regex pattern or hex string.
+    #[arg(long, value_name = "PATTERN")]
+    pub ends_with: Option<String>,
+
+    /// Return mnemonic with address (default false)
+    #[arg(long)]
+    pub mnemonic: bool,
+
+    /// Path to save the generated vanity address to.
+    #[arg(long, value_hint = clap::ValueHint::FilePath, value_name = "PATH")]
+    pub save_path: Option<PathBuf>,
+}
+
+pub fn handler(args: Arg) -> Result<serde_json::Value> {
+    let Arg {
+        starts_with,
+        ends_with,
+        mnemonic,
+        save_path,
+    } = args;
+
+    let mut left_exact_hex = None;
+    let mut left_regex = None;
+    if let Some(prefix) = starts_with {
+        match parse_pattern(&prefix, true)? {
+            Either::Left(left) => left_exact_hex = Some(left),
+            Either::Right(re) => left_regex = Some(re),
+        }
+    }
+
+    let mut right_exact_hex = None;
+    let mut right_regex = None;
+    if let Some(suffix) = ends_with {
+        match parse_pattern(&suffix, false)? {
+            Either::Left(right) => right_exact_hex = Some(right),
+            Either::Right(re) => right_regex = Some(re),
+        }
+    }
+
+    println!("Starting to generate vanity address...");
+    let start_time = Instant::now();
+
```

Comment:
If the user is using a regex, should we let them match whatever they're looking for rather than only applying it to the start or end? i.e should this instead handle:

- left
- right
- left and right
- regex

The reason I'm asking is that if they're using a regex, they could put the left or right or both IN the regex.

Reply:
That's correct. However, in most cases, users typically choose a prefix and/or suffix based on characters, as this is what wallet UIs commonly display.
For a better user experience, I believe it's best to offer prefix/suffix options, since users are less likely to remember regex syntax offhand.